developer_manual/general/security.rst

```diff
@@ -1,51 +1,408 @@
+If it needs data from these tables, it should call an API to retrieve it.
+
+Error Messages and Error Pages
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Don't show sensitive information on error pages or in error messages. 
+Sensitive information includes:
+
+- Username/password
+- E-Mail addresses
+- Version numbers
+- Paths
+
+Don't show information in error messages or on error pages, which is too detailed.
+
+**Example:**
+
+If a user can't login, don't show an error like: "*Your password is wrong*". 
+Instead, show a message such as: "*There was an error with your credentials*". 
+If you print "*Your password is wrong*" then an attacker knows the username was a valid one in the ownCloud installation.
+Additionally, consider implementing a `CAPTCHA`_ to prevent brute force attacks, after five failed login attempts.
+
+Session ID Transport
+~~~~~~~~~~~~~~~~~~~~
+
+Don't use a session id as a GET Parameter, because these persist in browser history.
+Use cookies instead.
+
+New Session ID After a Successful Login
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+After a successful login, the session id should be regenerated to prevent `session fixation attacks`_. 
+If you have to switch between HTTPS and HTTP, you should change the session id, because an attacker could have already read the session id.
+
+Access Protection With Authorization Check
+------------------------------------------
+
+Every request to the server must perform a check if the user has the authorization to perform this request. 
+A client-side check is not recommended. 
+However, it can improve the user's experience.
+
+Best Practices
+--------------
+
+Use of the eval Function 
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Don't use either PHP's or JavaScript's ``eval`` functions — especially not with user-supplied data.
+
+Input Validation
+~~~~~~~~~~~~~~~~
+
+All user-supplied data, ``$_SERVER``, and ``$_COOKIE`` variables **must** be validated. 
+All these contain data which can be changed (or forged) by the client.
+You should also sanitize any supplied script code. 
+
+**Example:**
+
+If you expect to receive an integer id as a GET parameter, then always explicitly cast it into an integer using the cast operator ``(int)``, because all ``$_REQUEST`` parameter are strings. 
+However, if you expect text as a parameter, use `PHP's htmlspecialchars function`_ with ``ENT_QUOTES`` or ``strip_tags`` to prevent `Cross-site Scripting (XSS) attacks`_.
+
+.. code-block:: php
+
+  <?php
+
+  $neu = htmlspecialchars("<a href='test'>Test</a>", ENT_QUOTES);
+  echo $neu; // &lt;a href=&#039;test&#039;&gt;Test&lt;/a&gt;
+
+.. code-block:: php 
+
+  <?php
+
+  $text = '<p>Test-Absatz.</p><!-- Kommentar --> <a href="#fragment">Anderer Text</a>';
+  echo strip_tags($text);
+  echo "\n";
+
+**Output:**
+
+.. code-block:: console 
+
+  Test-Absatz. Anderer Text
+  <p>Test-Absatz.</p> <a href="#fragment">Anderer Text</a>
+
+Please do the validation **before** all other actions.
```

Comment:
Don't need to be so polite here.
s/Please do/Do/

Reply:
That's a habit I'm gradually weaning myself off of.